wire-ios-data-model/Source/ConversationList/ConversationList.swift

```diff
@@ -0,0 +1,230 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import CoreData
+
+/// A collection of conversation instances with additional infos.
+@objc(ZMConversationList) @objcMembers
+public final class ConversationList: NSObject {
+
+    public weak var managedObjectContext: NSManagedObjectContext?
+
+    let identifier: String
+    let label: Label?
+    public private(set) var items: [ZMConversation]
+    private let conversationKeysAffectingSorting: NSSet
+    private var filteringPredicate: NSPredicate
+    private let sortDescriptors: [NSSortDescriptor]
+
+    public convenience init(
+        allConversations: [ZMConversation],
+        filteringPredicate: NSPredicate,
+        managedObjectContext: NSManagedObjectContext,
+        description: String
+    ) {
+        self.init(
+            allConversations: allConversations,
+            filteringPredicate: filteringPredicate,
+            managedObjectContext: managedObjectContext,
+            description: description,
+            label: nil
+        )
+    }
+
+    public init(
+        allConversations: [ZMConversation],
+        filteringPredicate: NSPredicate,
+        managedObjectContext: NSManagedObjectContext,
+        description: String,
+        label: Label?
+    ) {
+        self.filteringPredicate = filteringPredicate
+        self.managedObjectContext = managedObjectContext
+        identifier = description
+        self.label = label
+        sortDescriptors = ZMConversation.defaultSortDescriptors()!
+
+        conversationKeysAffectingSorting = Self.calculateKeysAffectingPredicateAndSort(sortDescriptors)
+        items = Self.createItems(allConversations, filteringPredicate: filteringPredicate)
+
+        super.init()
+
+        managedObjectContext.performAndWait {
+            managedObjectContext.conversationListObserverCenter.startObservingList(self)
+        }
+    }
+
+    deinit {
+        if let managedObjectContext {
+            managedObjectContext.performAndWait {
+                managedObjectContext.conversationListObserverCenter.removeConversationList(self)
+            }
+        }
+    }
+
+    private static func createItems(_ conversations: [ZMConversation], filteringPredicate: NSPredicate) -> [ZMConversation] {
+        let filtered = (conversations as NSArray).filtered(using: filteringPredicate)
+        return NSSet(array: filtered).sortedArray(using: ZMConversation.defaultSortDescriptors()!) as! [ZMConversation]
```

Comment:
Should we sort using the stored sort descriptors?

Reply:
Good idea, done: 1751a4545d4db6850545dd46203f34538098d131